### PR TITLE
Post: How to speed up inference in your deep learning model

### PR DESCRIPTION
**Pull request by renzodgc**
Did you obtain good accuracy with your deep learning model only to find out that inference time isn't enough to deploy on production? Are you lost on how to optimize your model's inference speed? Then this post is for you.

**TODO:**
- [x] Change date
- [x] Change featured image
- [x] Change crosspost to medium to true?

### Preview URL
- https://blog-staging.xmartlabs.com/2020/05/20/how-to-speed-up-inference-in-your-deep-learning-model/

### CODE REVIEW

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+Did you obtain good accuracy with your deep learning model only to find out that inference time isn't enough to deploy on production? Are you lost on how to optimize your model's inference speed? Then this post is for you.
+
+<div style="text-align: center"><img width="100%" src="/images/automate-manual-testing/featured2.png" /></div>
+
+Data Science projects have the peculiarity that your focus must constantly shift across completely different requirements.
+
+- How will you obtain the data? Are you creating the dataset? (if so what exactly should you label? how many samples should you invest in?)
+- How will you make use of the model? A model is nice and fine, but how do you make a product that adds value to the user?
+- What model will you implement? There are many scientific papers out there, and often from the time you begin research until you are in the middle of implementation, the state of the art has already changed.
+- But the most important question of all. Is all this viable? Can a model be trained using the data your budget allows? Will it perform sufficiently good for the use we intend the model to?
```
Comment (mlvieras) [inline _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:25]:
```suggestion
- But the most important question of all. Is all of this viable? Can a model be trained using the data your budget allows? Will it perform sufficiently good for the use case we intend to apply it to?
```

**Comment by bryant1410** [comment]
can you fix the staging deploy?

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+When we tried to quantize a PyTorch [Faster R-CNN](https://arxiv.org/abs/1506.01497) model we, unfortunately, run into multiple issues.
+
+At the time of writing this article, PyTorch doesn't natively support model quantization for GPUs [as TensorFlow does](https://www.tensorflow.org/api_docs/python/tf/quantization/quantize).
+
+So in order to quantize a PyTorch model, we must run it on [NVIDIA TensorRT](https://developer.nvidia.com/tensorrt), but for running it in TensorRT runtime we must convert our PyTorch model.
+As of now, the only way to achieve this conversion is to first convert our PyTorch model to [ONNX](https://github.com/onnx/onnx), and then convert it to TensorRT.
+
+On paper, all this makes sense and should be easy enough to do.
+However, in practice issues may occur from all these conversions.
+This is mainly due to the fact that the development of PyTorch, ONNX, and TensorRT goes in different ways.
```
Comment (mlvieras) [inline _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:66]: What does this sentence mean? That the three types of models are incompatible? It's not immediately clear to me what "development" means in this context.

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+This is mainly due to the fact that the development of PyTorch, ONNX, and TensorRT goes in different ways. 
+
+- You may be able to run the model on ONNX, but [issues may occur when converting ONNX to TensorRT](https://github.com/onnx/onnx-tensorrt/issues/302), especially with some layers, [such as the Resize layer in PyTorch](https://github.com/NVIDIA/TensorRT/issues/284).
+- At the time we tried to do this conversion, it happened to us that [models built with PyTorch v1.3 or higher can be converted and run in ONNX Runtime, but can't be run through the ONNX optimizer](https://github.com/onnx/onnx/issues/2417) (which makes important improvements in the converted network).
+
+Some users have managed to solve their issues during conversion by downgrading PyTorch.
+However, this limits the ONNX `opset` you can access, which in turn also limits what TensorRT version will you be able to run your engine with.
+
+Hopefully, all these issues will be solved in the near future...
+But considering that all these frameworks have high development speed it's likely that there always be short time incompatibilities.
```
Comment (mlvieras) [inline _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:75]:
```suggestion
But considering that all these frameworks have high development speed it's likely that there will always be short-lived incompatibility periods.
```

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+---
+
+### Converting the model to half-precision
+
+In a similar fashion to the previous approach, this alternative aims to trade-off accuracy for speed and memory efficiency.
+It offers a middle point between FP32 and UInt8, where:
+
+- The model size is reduced by up to half (instead of by up to 75%)
+- The difference of accuracy is negligible compared to FP32, this means not sacrificing as much accuracy as when applying model quantization.
```
Comment (mlvieras) [inline _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:88]: This sounds like quite the statement to me. Is there general evidence of this? I'd expect a link to one or more information sources that support this claim. I would phrase it differently: `We have experienced a negligible accuracy reduction in our experiments` or something similar.

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+This approach is called [Automatic Mixed Precision](https://arxiv.org/abs/1710.03740), and it differs on quantization in that rather than post-processing the weights of a trained model, the model should be trained, using mixed precision from the get-go.
+
+**Our experience**
+
+TensorFlow once again goes great extents to make our lives easier by [giving us native support to achieve this](https://www.tensorflow.org/api_docs/python/tf/train/experimental/enable_mixed_precision_graph_rewrite).
+Whereas if you're using PyTorch, [NVIDIA Apex](https://github.com/NVIDIA/apex) is the tool you should use, which fortunately according to our experiences has way fewer issues than the pain that is doing quantization with a PyTorch model.
+
+Integrating Apex is advertised as a mere addition of three lines to your code.
+And actually, it is not much more than that.
+You must initialize it, change the backward pass call to use Apex's scaled loss, and modify the way you save and load checkpoints.
```
Comment (mlvieras) [inline _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:105]: Would it make sense to have a code snippet here? Or is the documentation really clear?

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+If you want to know more about knowledge distillation check out this paper by [Geoffrey Hinton et al](https://arxiv.org/pdf/1503.02531.pdf).
+
+---
+
+## To sum up
+
+In this blog, we've described five approaches to improve the inference time of your deep learning model.
+In particular, we'd advise you to implement them in the order we also listed them in, because any coding we do to implement model quantization and automatic mixed-precision is of great value to any further changes we make on our model.
+
+We hope that this article was of value to you, either by giving you guidance on an issue you're currently facing or by arming you with our knowledge should the need arise for it!
```
Comment (mlvieras) [inline _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:175]: I would love a table here comparing the five approaches in accuracy, speed and coding difficulty.

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md (hunk 2 above)
Comment (renzodgc) [reply _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:65]: It refers to the development of the frameworks (each company goes in different directions).
Maybe:
```suggestion
This is mainly due to the fact that the development of PyTorch, ONNX, and TensorRT goes in multiple directions when a feature is added into one an old integration won't necessarily support it. 
```

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md (hunk 4 above)
Comment (renzodgc) [reply _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:87]: I'll rewrite and add some difference to fully converting to fp16 and doing automatic mixed precision

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md (hunk 5 above)
Comment (renzodgc) [reply _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:104]: We discussed it and it didn't add that much, considering the apex documentation is clear enough. But we could revisit the idea @mats-claassen, what do you think?

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md (hunk 6 above)
Comment (renzodgc) [reply _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md:174]: Not so sure if we are able to provide such a thing.

The first two approaches can either solve your issue (if it needs only a small push) or add value to the following approaches.

The increment in accuracy and speed will vary on your problems and the models you use. Maybe you were already using the best model for your use case. In a similar way, the coding difficulty could be from almost none (if you found an open-source that implements a model) to a lot if you have to do it yourself...

Regarding the 5th approach, we didn't implement it, so we can't exactly know the difficulties there, aside from the fact that you would need to first have a really accurate model (or reuse previous trained model as an ensemble)

Comment (renzodgc) [reply _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md]: @mats-claassen do you think there is more insight we can add regarding that end?

**Comment by renzodgc** [comment]
> can you fix the staging deploy?

 @bryant1410 it is now fixed

File: _posts/2020-06-01-how-to-speed-up-inference-in-your-deep-learning-model.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
+---
+layout: post
+title: "How to speed up inference in your deep learning model"
```
Comment (bryant1410) [inline _posts/2020-06-01-how-to-speed-up-inference-in-your-deep-learning-model.md:3]: What about:

```suggestion
title: "5 Practical Ways to Speed Up your Deep Learning Model"
```

Comment (renzodgc) [reply _posts/2020-06-01-how-to-speed-up-inference-in-your-deep-learning-model.md]: I like it :)

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md (hunk 5 above)
Comment (mats-claassen) [reply _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md]: The doc is very clear, but as it is only 4 lines we could add it here

File: _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md (hunk 6 above)
Comment (mats-claassen) [reply _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md]: There are no numbers we could give for that comparison. It heavily depends on the model and the idea of the blog is to be generic in that aspect, right? What could be done is a table with String values like 'High', 'Medium', 'Low'. So we could say accuracy 'Slightly decreases', speed 'Faster', coding difficulty 'Medium'. But without numbers I think the usefulness of that is limited. What do you think?

Comment (renzodgc) [reply _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md]: I don't see much value in it.
If those cases don't apply to most general cases then the information will be mostly not applicable.
My idea was that the article should convey to the viewer what are the difficulties and possible benefits of applying an approach to a case they have in mind.

Comment (renzodgc) [reply _posts/2020-05-20-how-to-speed-up-inference-in-your-deep-learning-model.md]: Like with quantization and AMP the strengths and difficulties are clear. However, with changes in the model it heavily depends on:
- The implementation's reusability
- If they already implemented the best model in theory for their use case

File: _posts/2020-06-01-how-to-speed-up-inference-in-your-deep-learning-model.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+**Our experience**
+
+TensorFlow once again goes great extents to make our lives easier by [giving us native support to achieve this](https://www.tensorflow.org/api_docs/python/tf/train/experimental/enable_mixed_precision_graph_rewrite).
+Whereas if you're using PyTorch, [NVIDIA Apex](https://github.com/NVIDIA/apex) is the tool you should use, which fortunately according to our experiences has way fewer issues than the pain that is doing quantization with a PyTorch model.
+
+Integrating Apex is advertised as a mere addition of three lines to your code.
+And actually, it is not much more than that.
+You must initialize it, change the backward pass call to use Apex's scaled loss, and modify the way you save and load checkpoints.
+The results are pretty good as well.
```
Comment (mats-claassen) [inline _posts/2020-06-01-how-to-speed-up-inference-in-your-deep-learning-model.md:106]:
```suggestion
```python
# Initialization
opt_level = 'O1'
model, optimizer = amp.initialize(model, optimizer, opt_level=opt_level)

# Train your model
...
with amp.scale_loss(loss, optimizer) as scaled_loss:
    scaled_loss.backward()
...
```
The results are pretty good as well.
```
